lib/THC/THCTensorIndex.cu

```diff
@@ -257,78 +257,87 @@ void THCudaTensor_indexFill(THCState *state, THCudaTensor *res_, int dim, THCuda
-      const long srcIdx = ((long) index[idx] - 1) * stride;
-      const long targetIdx = idx * stride;
-
-      #pragma unroll
-      for (int i = 0; i < VT; i++) {
-        const int featureIdx = startIdx + i * WARP_SIZE;
-        if (featureIdx < stride) {
-          tensor[targetIdx + featureIdx] = src[srcIdx + featureIdx];
-        }
-      }
+// We prefer this kernel to avoid reloading index points if the number
+// of indices per each slice is a small number.
+// This kernel in fact works for all choices of problem size, but if
+// the number of indices chosen is large, then the
+// indexSelectLargeIndex kernel is a better choice to increase
+// parallelism.
+template <typename IndexType, int DstDim, int SrcDim, int IdxDim>
+__global__ void indexSelectSmallIndex(TensorInfo<IndexType> dst,
+                                      TensorInfo<IndexType> src,
+                                      TensorInfo<IndexType> indices,
+                                      int dstSelectDim,
+                                      int srcSelectDim,
+                                      IndexType totalSize,
+                                      IndexType innerSize) {
+  // In order to avoid reloading the index that we are copying, load
+  // it once per thread per slice. This kernel is chosen when this is
+  // a good choice, since re-accessing indices can be slow.
+  for (IndexType dstIndex = 0; dstIndex < indices.sizes[0]; ++dstIndex) {
+    IndexType srcIndex =
+      indices.data[IndexToOffset<IndexType, IdxDim>::get(dstIndex, indices)] - 1;
+
+    // We stride over the output ignoring the indexed dimension
+    // (innerSize), whose offset calculation is handled differently
+    for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+         linearIndex < innerSize;
+         linearIndex += gridDim.x * blockDim.x) {
+      IndexType dstOffset =
+        IndexToOffset<IndexType, DstDim>::get(linearIndex, dst);
+      dstOffset += dstIndex * dst.strides[dstSelectDim];
+
+      IndexType srcOffset =
+        IndexToOffset<IndexType, SrcDim>::get(linearIndex, src);
+      srcOffset += srcIndex * src.strides[srcSelectDim];
+
+      dst.data[dstOffset] = src.data[srcOffset];
     }
   }
 }
 
-__global__ void THCudaTensor_kernel_indexSelect(
-  TensorInfo<unsigned long> dest,
-  TensorInfo<unsigned long> src,
-  TensorInfo<unsigned long> indices,
-  long totalSize,
-  int indexDim
-)
-{
-  for (unsigned long linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+// We prefer this kernel to balance parallelism across index points
+// and slices otherwise.
+// This kernel in fact works for all choices of problem size, but if
+// the number of indices chosen is small, then the
+// indexSelectSmallIndex kernel is a better choice to reduce memory
+// accesses.
+template <typename IndexType, int DstDim, int SrcDim, int IdxDim, bool IndexInnermost>
```

Comment:
Yes, I added this as an experiment. I was thinking about specializing what the index goes over innermost for better performance based on some heuristic, but I'll just remove it.